datafusion/optimizer/src/simplify_expressions/guarantees.rs

```diff
@@ -0,0 +1,514 @@
+                    _ => return Ok(expr),
+                };
+
+                // Check if this is a comparison between a column and literal
+                let (col, op, value) = match (left.as_ref(), right.as_ref()) {
+                    (Expr::Column(_), Expr::Literal(value)) => (left, *op, value),
+                    (Expr::Literal(value), Expr::Column(_)) => {
+                        // If we can swap the op, we can simplify the expression
+                        if let Some(op) = op.swap() {
+                            (right, op, value)
+                        } else {
+                            return Ok(expr);
+                        }
+                    }
+                    _ => return Ok(expr),
+                };
+
+                if let Some(col_interval) = self.guarantees.get(col.as_ref()) {
+                    let result = col_interval.apply_operator(&op, &value.into())?;
+                    if result.is_certainly_true() {
+                        Ok(lit(true))
+                    } else if result.is_certainly_false() {
+                        Ok(lit(false))
+                    } else {
+                        Ok(expr)
+                    }
+                } else {
+                    Ok(expr)
+                }
+            }
+
+            // Columns (if interval is collapsed to a single value)
+            Expr::Column(_) => {
+                if let Some(col_interval) = self.guarantees.get(&expr) {
+                    if let Some(value) = col_interval.single_value() {
+                        Ok(lit(value))
+                    } else {
+                        Ok(expr)
+                    }
+                } else {
+                    Ok(expr)
+                }
+            }
+
+            Expr::InList(InList {
+                expr: inner,
+                list,
+                negated,
+            }) => {
+                if let Some(interval) = self.guarantees.get(inner.as_ref()) {
+                    // Can remove items from the list that don't match the guarantee
+                    let new_list: Vec<Expr> = list
+                        .iter()
+                        .filter_map(|expr| {
+                            if let Expr::Literal(item) = expr {
+                                match interval.contains(&NullableInterval::from(item)) {
+                                    // If we know for certain the value isn't in the column's interval,
+                                    // we can skip checking it.
+                                    Ok(NullableInterval::NotNull { values })
+                                        if values == Interval::CERTAINLY_FALSE =>
+                                    {
+                                        None
+                                    }
+                                    Err(err) => Some(Err(err)),
+                                    _ => Some(Ok(expr.clone())),
+                                }
```

Comment:
Perhaps you can rewrite this using `is_certainly_false` like the following (not tested)

```suggestion
                                if interval.contains(&NullableInterval::from(item))?
                                    // If we know for certain the value isn't in the column's interval,
                                    // we can skip checking it.
                                  .is_certainly_false() {
                                        None
                                    }
                                    else {
                                      Some(Ok(expr.clone())),
                                }
```

Reply:
This is one of those awkward closures where we are returning `Option<Result<_>>`, so early returning result doesn't quite work. But you are right it can be simplified. 👍